File: search/binary_search.rb
# binary search in Ruby


def find_number_index(array, y)
  left = 0
  right = array.size - 1

  while(right >= left)
    mid = left + (right - left) / 2

    return mid if array[mid] == y
    left = mid + 1 if array[mid] < y
    right = mid - 1 if array[mid] > y
  end

  return -1

end

array = [8, 9, 15, 18, 19, 24, 29, 31, 33, 36, 43, 50, 99, 101]
y = 101

result = find_number_index(array, y)

p "#{y}はarrayの#{result + 1}番目にあります" if result > 0
p "#{y}はarrayに存在しません" if result < 0

File: search/binary_search.c
/* binary search in C */

#include <stdio.h>
#define N 14

int find_number_index(int *array, int y);

int main(void) {
  int array[] = {8, 9, 15, 18, 19, 24, 29, 31, 33, 36, 43, 50, 99, 101};
  int y = 36;
  int result = find_number_index(array, y);

  if (result > 0)
    printf("%dはarrayの%d番目にあります", y, result + 1);
  else
    printf("%dはarrayに存在しませ", y);
  return 0;
}

int find_number_index(int *array, int y) {
  int left = 0;
  int right = N - 1;

  while(right >= left) {
    int mid = left + (right - left) / 2;

    if (array[mid] == y) {
      return mid;
    }

    if (array[mid] < y) {
      left = mid + 1;
    }

    if (array[mid] > y) {
      right = mid - 1;
    }
  }

  return -1;
}

